### PR TITLE
modMcf account package unit tests

### PR DESCRIPTION
**Pull request by aionWilliam**
## Notice

It is not allowed to submit your PR to the master branch directly, please submit your PR to the master-pre-merge branch.

## Description

Please include a brief summary of the change that this pull request proposes. Include any relevant motivation and context. List any dependencies required for this change.

-Created unit test coverage for package org.aion.mcf.account in modMcf.

Fixes Issue # .

## Type of change

Insert **x** into the following checkboxes to confirm (eg. [x]):
- [ ] Bug fix.
- [ ] New feature.
- [ ] Enhancement.
- [x] Unit test.
- [ ] Breaking change (a fix or feature that causes existing functionality to not work as expected).
- [ ] Requires documentation update.

## Testing

Please describe the tests you used to validate this pull request. Provide any relevant details for test configurations as well as any instructions to reproduce these results.

-AccountManagerTest (new)
-AccountTest (new)
-FileDateTimeComparatorTest  (updated)
-KeystoreTest (updated)

## Verification

Insert **x** into the following checkboxes to confirm (eg. [x]):
- [x] I have self-reviewed my own code and conformed to the style guidelines of this project.
- [x] New and existing tests pass locally with my changes.
- [x] I have added tests for my fix or feature.
- [ ] I have made appropriate changes to the corresponding documentation.
- [x] My code generates no new warnings.
- [x] Any dependent changes have been made.

### CODE REVIEW

File: modMcf/test/org/aion/mcf/account/AccountManagerTest.java [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+        assertTrue(accountManager.unlockAccount(Address.wrap(k2.getAddress()), p2, 1));
+
+        // time out more than max
+        assertTrue(accountManager.unlockAccount(Address.wrap(k3.getAddress()), p3, 86401));
+
+        // not registered key
+        assertFalse(accountManager.unlockAccount(address1, "no pass", 2000));
+
+        // account already present, update the timeout
+        assertTrue(accountManager.unlockAccount(Address.wrap(k1.getAddress()), p1, 4000));
```
Comment (aion-kelvin) [inline modMcf/test/org/aion/mcf/account/AccountManagerTest.java]: AccountManager keeps a Map from address to account.  For unlockAccount, if the given address is not in the map, it adds it into it and creates a new account object.  The values of the map are accessible through getAccounts().  As part of verifying unlockAccount, can you look at the results in getAccounts() to make sure it's consistent with what you'd expect after calling unlockAccount?  Also, check that the Account object in it has the correct key and timeout.

File: modMcf/test/org/aion/mcf/account/AccountManagerTest.java [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+
+        // lock some accounts ------------------------------------------------------------------------------------------
+        assertTrue(accountManager.lockAccount(Address.wrap(k2.getAddress()), p2));
```
Comment (aion-kelvin) [inline modMcf/test/org/aion/mcf/account/AccountManagerTest.java]: lockAccount modifies the Account object that is internally stored in AccountManager.  You can access it through getAccounts(); as part of verifying lockAccount, can you see that the Account object is updated correctly (I think the timeout value is changed when you lock it).

File: modMcf/test/org/aion/mcf/account/AccountManagerTest.java [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+    public void testSingletonAccountManager(){
+        // first check that there are no accounts
+        assertEquals(0, accountManager.getAccounts().size());
+
+        // unlock some accounts ----------------------------------------------------------------------------------------
+        assertTrue(accountManager.unlockAccount(Address.wrap(k1.getAddress()), p1, 2000));
+        assertTrue(accountManager.unlockAccount(Address.wrap(k2.getAddress()), p2, 1));
+
+        // time out more than max
+        assertTrue(accountManager.unlockAccount(Address.wrap(k3.getAddress()), p3, 86401));
```
Comment (arajasek) [inline modMcf/test/org/aion/mcf/account/AccountManagerTest.java]: This test would be more meaningful if you asserted that the account's new timeout was, in fact, less than the max value.

File: modMcf/test/org/aion/mcf/account/KeystoreTest.java [hunk 4]
```diff
@@ -195,4 +193,43 @@ public void testList() {
...
+
+        importKey.put(key.toString(), key.toString());
+        Set<String> res = Keystore.importAccount(importKey);
+    }
+
+    @Test
+    public void testAccountSorted(){
+        List<String> res = Keystore.accountsSorted();
+        for(String addr: res)
+            System.out.println(addr);
```
Comment (aion-kelvin) [inline modMcf/test/org/aion/mcf/account/KeystoreTest.java]: I think you'll need to assert something here

File: modMcf/test/org/aion/mcf/account/KeystoreTest.java [hunk 5]
```diff
@@ -195,4 +193,43 @@ public void testList() {
...
+        Keystore.importAccount(null);
+    }
+
+    @Test
+    public void testImportAccount(){
+        Map<String, String> importKey = new HashMap<>();
+        ECKey key = ECKeyFac.inst().create();
+
+        importKey.put(key.toString(), key.toString());
+        Set<String> res = Keystore.importAccount(importKey);
```
Comment (aion-kelvin) [inline modMcf/test/org/aion/mcf/account/KeystoreTest.java]: Nothing gets asserted/verified here... did you forget to add them?

Also, right now when I run the test it doesn't go into the if-block at Keystore line 307 -- it just fails the if check and then the method exits.  I'm not sure if that is intentional but it seems to be skipping an important part of the code it is supposed to test.

File: modMcf/test/org/aion/mcf/account/KeystoreTest.java [hunk 6]
```diff
@@ -99,6 +96,7 @@ public void keyCreateAndRetrieve2() {
     @Test
     public void testKeyCreate() {
         String password = randomPassword();
+
```
Comment (aion-kelvin) [inline modMcf/test/org/aion/mcf/account/KeystoreTest.java]: My comments on AccountManager about temp file and unit vs. integ test apply here also.  I know you didn't create this test originally, but can you try to figure out a way to clean up all the files created by this test?  I think every time the Keystore.create method gets hit, a file gets created on the file system.

File: modMcf/test/org/aion/mcf/account/AccountManagerTest.java [hunk 7]
```diff
@@ -0,0 +1,111 @@
...
+
+        // account already present, update the timeout
+        assertTrue(accountManager.unlockAccount(Address.wrap(k1.getAddress()), p1, 4000));
+
+        // lock some accounts ------------------------------------------------------------------------------------------
+        assertTrue(accountManager.lockAccount(Address.wrap(k2.getAddress()), p2));
+
+        // not registered key
+        assertFalse(accountManager.lockAccount(address1, "no pass"));
+
```
Comment (arajasek) [inline modMcf/test/org/aion/mcf/account/AccountManagerTest.java]: I would throw in lock and unlock tests when they're called with the wrong password as well.

File: modMcf/test/org/aion/mcf/account/AccountManagerTest.java [hunk 8]
```diff
@@ -0,0 +1,111 @@
+/*
+ ******************************************************************************
+ * Copyright (c) 2017-2018 Aion foundation.
```
Comment (arajasek) [inline modMcf/test/org/aion/mcf/account/AccountManagerTest.java]: I think this should use this header instead:

https://github.com/aionnetwork/aion/wiki/Aion-License-Header#2-new-functionality-developed-by-the-aion-foundation-will-have-the-license-below

File: modMcf/test/org/aion/mcf/account/AccountManagerTest.java [hunk 9]
```diff
@@ -0,0 +1,111 @@
...
+        k3 = ECKeyFac.inst().create();
+
+        Keystore.create(p1, k1);
+        Keystore.create(p2, k2);
+        Keystore.create(p3, k3);
+    }
+
+    @Test
+    public void testSingletonAccountManager(){
+        // first check that there are no accounts
```
[the remark this answers is not in the document]
Comment (arajasek) [reply modMcf/test/org/aion/mcf/account/AccountManagerTest.java]: I agree that refactoring this into separate tests is advisable.

File: modMcf/test/org/aion/mcf/account/AccountManagerTest.java [hunk 10]
```diff
@@ -0,0 +1,111 @@
...
+    private AccountManager accountManager = AccountManager.inst();
+    private Address address1 = Address.wrap("a011111111111111111111111111111101010101010101010101010101010101");
+
+    @Before
+    public void setup(){
+        k1 = ECKeyFac.inst().create();
+        k2 = ECKeyFac.inst().create();
+        k3 = ECKeyFac.inst().create();
+
+        Keystore.create(p1, k1);
```
[the remark this answers is not in the document]
Comment (aionWilliam) [reply modMcf/test/org/aion/mcf/account/AccountManagerTest.java]: Thanks for the insightful review! I will try to tackle these issues one by one. First I will create a helper function that will clean up the tester keystore files in the keystore folder of modMcf.

File: modMcf/test/org/aion/mcf/account/FileDateTimeComparatorTest.java [hunk 11]
```diff
@@ -133,4 +134,29 @@ public void testComparatorWithGarbage() {
             }
         }
     }
+
+    @Test
+    public void testNullInput(){
+        File testFile = new File("test");
```
Comment (AionJayT) [inline modMcf/test/org/aion/mcf/account/FileDateTimeComparatorTest.java]: delete file after test

File: modMcf/test/org/aion/mcf/account/AccountManagerTest.java (hunk 3 above)
Comment (aionick) [inline modMcf/test/org/aion/mcf/account/AccountManagerTest.java]: Maybe worth making `UNLOCK_MAX` public in `AccountManager` since it is static final and immutable, then your test doesn't have to rely on a hard-coded value

File: modMcf/test/org/aion/mcf/account/AccountManagerTest.java [hunk 12]
```diff
@@ -0,0 +1,111 @@
...
+        Keystore.create(p2, k2);
+        Keystore.create(p3, k3);
+    }
+
+    @Test
+    public void testSingletonAccountManager(){
```
Comment (aionick) [inline modMcf/test/org/aion/mcf/account/AccountManagerTest.java]: Could you also test the functionality of unlocking & locking the same account multiple times?

File: modMcf/test/org/aion/mcf/account/FileDateTimeComparatorTest.java [hunk 13]
```diff
@@ -133,4 +134,29 @@ public void testComparatorWithGarbage() {
...
+    @Test
+    public void testInvalidName(){
+        File testFile = new File("test");
+        FileDateTimeComparator fileDateTimeComparator = new FileDateTimeComparator();
+
+        assertEquals(0, fileDateTimeComparator.compare(testFile,testFile));
+    }
+
+    @Test
+    public void testNullTime() {
```
Comment (aionick) [inline modMcf/test/org/aion/mcf/account/FileDateTimeComparatorTest.java]: can you use a better name here since the time isn't `null` per se

File: modMcf/test/org/aion/mcf/account/FileDateTimeComparatorTest.java [hunk 14]
```diff
@@ -133,4 +134,29 @@ public void testComparatorWithGarbage() {
             }
         }
     }
+
+    @Test
```
Comment (aionick) [inline modMcf/test/org/aion/mcf/account/FileDateTimeComparatorTest.java]: I would like to see more edge case tests. Timestamps, especially a non-trivial one like this, can be prone to small errors. See what happens when you give a month with 3 digits, or two times that differ by the absolute smallest quantity the timestamp allows for, etc.

File: modMcf/test/org/aion/mcf/account/KeystoreTest.java [hunk 15]
```diff
@@ -195,4 +193,43 @@ public void testList() {
 
         assertTrue(hasAddr);
     }
+
+    @Test (expected = NullPointerException.class)
+    public void testBackupAccountWithNullInput(){
+        Keystore.backupAccount(null);
+    }
+
```
Comment (aionick) [inline modMcf/test/org/aion/mcf/account/KeystoreTest.java]: The `list` functionality could be tested better. I know you didn't write that test but its for-loop doesn't make much sense (testing if the list is size 1 and contains an address). But it would be good to also test this when you expect no entries and when you expect multiple entries.

File: modMcf/test/org/aion/mcf/account/KeystoreTest.java (hunk 6 above)
Comment (aionWilliam) [reply modMcf/test/org/aion/mcf/account/KeystoreTest.java]: sure, ill check where files are created and make sure they are removed after the test functions.

File: modMcf/test/org/aion/mcf/account/KeystoreTest.java [hunk 16]
```diff
@@ -79,6 +74,7 @@ public void keyCreateAndRetrieve() {
         System.out.println("new addr: " + address);
         ECKey key = Keystore.getKey(address, password);
         assertNotNull(key);
+        cleanFiles(address);
```
Comment (AionJayT) [inline modMcf/test/org/aion/mcf/account/KeystoreTest.java]: you can add @After teardown method() { delete files } to do the file delete.

File: modMcf/test/org/aion/mcf/account/AccountManagerTest.java [hunk 17]
```diff
@@ -0,0 +1,320 @@
...
+
+        for(File file: AllFilesInDirectory){
+            allFileNames.add(file.getName());
+        }
+
+        // get a list of the files needed to be deleted, check the ending of file names with corresponding addresses
+        for(String name: allFileNames){
+            String ending = name.substring(name.length()-64);
+
+            if(ending.equals(address1) || ending.equals(address2) || ending.equals(address3)) {
```
Comment (aion-kelvin) [inline modMcf/test/org/aion/mcf/account/AccountManagerTest.java]: Thanks for taking the time to ensure you're only deleting files actually created by your test

File: modMcf/src/org/aion/mcf/account/FileDateTimeComparator.java [hunk 18]
```diff
@@ -76,6 +77,16 @@ public int compare(File arg0, File arg1) {
 
         // TODO: make sure this is the same as the pattern we use to encode
```
Comment (AlexandraRoatis) [inline modMcf/src/org/aion/mcf/account/FileDateTimeComparator.java]: Is this TODO solved? If so, please remove/edit it.

Comment (aionWilliam) [reply modMcf/src/org/aion/mcf/account/FileDateTimeComparator.java]: There should be no issues, I will remove it
